cli/src/main/scala/com/codacy/analysis/cli/Main.scala

```diff
@@ -49,18 +49,44 @@ class MainImpl extends CLIApp {
         val commitUuid: Option[Commit.Uuid] =
           analyse.commitUuid.orElse(Git.currentCommitUuid(projectDirectory))
 
-        val analysisResults = analysis(analyse, projectDirectory, codacyClientOpt)
-
-        val uploadResult = upload(analyse, commitUuid, codacyClientOpt, analysisResults)
-
-        new ExitStatus(analyse.maxAllowedIssues, analyse.failIfIncompleteValue).exitCode(analysisResults, uploadResult)
+        //TODO(31/08/2018): In the next tickets:
+        // (1) configs will be converted to common model and passed through parameter to validate
+        // (2) validate commit cli parameter with commit retrieved from jGit
+        val analysisAndUpload = for {
+          _ <- validate(projectDirectory, analyse.uploadValue)
+          analysisResults <- analysis(analyse, projectDirectory, codacyClientOpt)
+          _ <- upload(analyse, commitUuid, codacyClientOpt, analysisResults)
+        } yield {
+          analysisResults
+        }
+
+        new ExitStatus(analyse.maxAllowedIssues).exitCode(analysisAndUpload)
     }
   }
 
-  private def analysis(
-    analyse: Analyse,
-    projectDirectory: File,
-    codacyClientOpt: Option[CodacyClient]): Either[AnalyseExecutor.ErrorMessage, Seq[ExecutorResult]] = {
+  private def validate(directory: File, upload: Boolean): Either[CLIErrorMessage, Unit] = {
+    (for {
+      repo <- Git.repository(directory)
+      uncommitedFiles <- repo.uncommitedFiles
+    } yield {
+      if (uncommitedFiles.nonEmpty) {
+        val error = CLIErrorMessage.UncommitedChanges(uncommitedFiles)
+        if (upload) {
+          logger.error(error.message)
+          Left(error)
+        } else {
+          logger.warn(CLIErrorMessage.UncommitedChanges(uncommitedFiles).message)
+          Right(())
+        }
+      } else {
+        Right(())
+      }
+    }).getOrElse(Right(()))
```

Comment:
The for comprehension is already made of eithers and you're doing some logic that I believe ends up being equivalent to another flatMap. Shouldn't this just be another <- in the for? Maybe wrap it inside a function?